src/project_model/_unit_tests_discovery.rs

```diff
@@ -0,0 +1,139 @@
+//!
+//! Moreover, this approach gives us the information whether tests are ignored or not.
+//! Not achieved with 'cargo metadata'.
+//!
+//! **Alternative next steps:** Analyze how cargo discovers tests and compile only
+//! to the moment when unit tests can be obtained.
+
+use crate::bsp_types::RustBuildTargetData::Rust;
+use crate::bsp_types::{
+    BuildTarget, BuildTargetCapabilities, BuildTargetIdentifier, RustBuildTarget, RUST_ID,
+};
+use cargo_metadata::camino::Utf8PathBuf;
+use cargo_metadata::Edition;
+use serde::{Deserialize, Serialize};
+use std::io::{BufRead, BufReader};
+use std::process::{Command, Stdio};
+
+/// Struct for deserialization of test JSON from
+/// 'cargo +nightly test -- --list --format json -Zunstable-options' output
+///
+/// Deserialization truncates discovered test JSON as below
+///
+/// ```json
+/// {
+///    "type": "test",
+///    "event": "discovered",
+///    "name": "bsp_types::notifications::exit_build::tests::exit_build_method",
+///    "ignore": false,
+///    "ignore_message": "",
+///    "source_path": "src/bsp_types/notifications/exit_build.rs",
+///    "start_line": 18,
+///    "start_col": 8,
+///    "end_line": 18,
+///    "end_col": 25
+/// }
+///
+/// CargoTestTarget {
+///     name: "bsp_types::notifications::exit_build::tests::exit_build_method",
+///     ignore: false,
+///     ignore_message: "",
+///     source_path: "src/bsp_types/notifications/exit_build.rs",
+/// }
+/// ```
+#[derive(Debug, Serialize, Deserialize)]
+struct CargoTestTarget {
+    /// name of the test
+    pub name: String,
+    /// path to the file, where test is defined
+    pub source_path: Utf8PathBuf,
+    /// whether this test is marked with #[ignore] and should not be compiled
+    ignore: bool,
+    /// displayed message for ignored test
+    ignore_message: String,
+}
+
+impl From<&CargoTestTarget> for BuildTarget {
+    /// **Unfinished** implementation of mapping CargoTestTarget to BuildTarget
+    fn from(test_target: &CargoTestTarget) -> Self {
+        let rust_specific_data = Rust(RustBuildTarget {
+            edition: Edition::E2021,
+            required_features: vec![],
+        });
+
+        let mut base_directory = test_target.source_path.clone();
+        // we assume that cargo test returns valid path to file, which additionally has a parent
+        base_directory.pop();
+
+        BuildTarget {
+            id: BuildTargetIdentifier {
+                uri: format!("{}:{}", test_target.source_path, test_target.name),
+            },
+            display_name: Some(test_target.name.clone()),
+            base_directory: Some(format!("file://{}", base_directory)),
+            tags: vec![],
+            capabilities: BuildTargetCapabilities::default(),
+            language_ids: vec![RUST_ID.to_string()],
+            dependencies: vec![],
+            data: Some(rust_specific_data),
+        }
```

Comment:
ten kod już gdzieś widziałem :)

Reply:
Trochę to zmieniłem, wyodrębniając część logiki. Najlepiej byłoby stworzyć trait new dla BuildTarget ale to jest kwestia związana z wyodrębnianiem Crate'a BSP types - jeśli to zrobimy nie będziemy mogli definiować traitów. Natomiast definiowanie traitów pod nasz projekt w nowym crate, który ma być stricte związany z typami, a nie naszym projektem nie wydaje się być najlepszym pomysłem. Dlatego też usunąłem definicję traita From